scapy/layers/ppp.py

```diff
@@ -261,31 +261,43 @@ class DIR_PPP(Packet):
     fields_desc = [ByteEnumField("direction", 0, ["received", "sent"])]
 
 
+class _PPPProtoField(EnumField):
+    """
+    A field that can be either Byte or Short, depending on the PPP RFC.
+
+    See RFC 1661 section 2
+    <https://tools.ietf.org/html/rfc1661#section-2>
+    """
+    def getfield(self, pkt, s):
+        if ord(s[:1]) & 0x01:
+            self.fmt = "!B"
+            self.sz = 1
+        else:
+            self.fmt = "!H"
+            self.sz = 2
+        return super(EnumField, self).getfield(pkt, s)
+
+    def addfield(self, pkt, s, val):
+        if val < 0x100:
+            self.fmt = "!B"
+            self.sz = 1
+        else:
+            self.fmt = "!H"
+            self.sz = 2
+        return super(EnumField, self).addfield(pkt, s, val)
+
+
 class PPP(Packet):
     name = "PPP Link Layer"
-    fields_desc = [ShortEnumField("proto", 0x0021, _PPP_PROTOCOLS)]
+    fields_desc = [_PPPProtoField("proto", 0x0021, _PPP_PROTOCOLS)]
 
     @classmethod
     def dispatch_hook(cls, _pkt=None, *args, **kargs):
-        if _pkt:
-            first_byte = orb(_pkt[0])
-            if first_byte == 0xff:
-                return HDLC
-            # See RFC 1661 section 2
-            # <https://tools.ietf.org/html/rfc1661#section-2>
-            if first_byte & 0x01:
-                return PPP_
+        if _pkt and _pkt[:1] == b'\xff':
+            return HDLC
         return cls
 
 
-class PPP_(PPP):
-    fields_desc = [
-        ByteEnumField("proto", 0x21,
-                      {k: v for k, v in six.iteritems(_PPP_PROTOCOLS)
-                       if k < 0x100}),
-    ]
```

Comment:
This packet would crash the renderer. I eventually took the liberty of recoding the field